File: frontend_apps/auto_ledger/lib/services/auth_service.dart
import 'package:dio/dio.dart';
import 'api_service.dart';
import '../utils/secure_storage.dart';

class AuthService {
  static String _errorMessage(
    DioException error,
    String fallback,
  ) {
    final data = error.response?.data;

    if (data is Map<String, dynamic>) {
      final message = data['message'];

      if (message is List && message.isNotEmpty) {
        return message.first.toString();
      }

      if (message != null && message.toString().trim().isNotEmpty) {
        return message.toString();
      }
    }

    return fallback;
  }

  static Future<Map<String, dynamic>> registerUser(
    Map<String, dynamic> data,
  ) async {
    try {
      final response =
          await ApiService.dio.post('/auth/user/register', data: data);

      if (response.statusCode == 201 || response.statusCode == 200) {
        return {
          'success': true,
          'message': response.data is Map<String, dynamic>
              ? response.data['message'] ?? 'OTP sent successfully.'
              : 'OTP sent successfully.',
        };
      }

      return {
        'success': false,
        'message': 'Registration failed. Please try again.',
      };
    } on DioException catch (error) {
      return {
        'success': false,
        'message': _errorMessage(
          error,
          'Registration failed. Please check your NIC and email.',
        ),
      };
    } catch (e) {
      return {
        'success': false,
        'message': 'Registration failed. Please try again.',
      };
    }
  }

  static Future<Map<String, dynamic>> verifyRegistration(
    String nicNo,
    String otp,
  ) async {
    try {
      final response = await ApiService.dio.post(
        '/auth/user/verify-registration',
        data: {'nicNo': nicNo, 'otp': otp},
      );

      if (response.statusCode == 200 || response.statusCode == 201) {
        await SecureStorage.deleteToken();

        return {
          'success': true,
          'message': response.data is Map<String, dynamic>
              ? response.data['message'] ??
                  'Registration successful. Please login.'
              : 'Registration successful. Please login.',
        };
      }

      return {
        'success': false,
        'message': 'OTP verification failed.',
      };
    } on DioException catch (error) {
      return {
        'success': false,
        'message': _errorMessage(
          error,
          'Invalid OTP. Please try again.',
        ),
      };
    } catch (e) {
      return {
        'success': false,
        'message': 'OTP verification failed.',
      };
    }
  }

  static Future<bool> resendRegistrationOtp(String nicNo) async {
    try {
      final response = await ApiService.dio.post(
        '/auth/user/resend-registration-otp',
        data: {'nicNo': nicNo},
      );
      return response.statusCode == 200 || response.statusCode == 201;
    } on DioException {
      return false;
    } catch (e) {
      return false;
    }
  }

  static Future<Map<String, dynamic>> loginUser(
      String nicNo, String password, String deviceId) async {
    try {
      final response = await ApiService.dio.post(
        '/auth/user/login',
        data: {
          'nicNo': nicNo,
          'password': password,
          'deviceId': deviceId,
        },
      );
      if (response.statusCode == 200 || response.statusCode == 201) {
        final token = response.data['accessToken'];
        await SecureStorage.saveToken(token);
        return {'success': true};
      }
      return {'success': false};
    } on DioException catch (e) {
      if (e.response?.statusCode == 403 &&
          e.response?.data['code'] == 'DEVICE_MISMATCH') {
        return {
          'success': false,
          'isDeviceMismatch': true,
          'email': e.response?.data['email'] ?? '',
        };
      }
      return {'success': false};
    } catch (e) {
      return {'success': false};
    }
  }

  static Future<bool> forgotPasswordCheck(String nicNo, String email) async {
    try {
      final response = await ApiService.dio.post(
        '/auth/user/forgot-password-check',
        data: {
          'nicNo': nicNo,
          'email': email,
        },
      );
      return response.statusCode == 200 || response.statusCode == 201;
    } on DioException {
      return false;
    } catch (e) {
      return false;
    }
  }

  static Future<bool> verifyResetOtp(String nicNo, String email, String otp) async {
    try {
      final response = await ApiService.dio.post(
        '/auth/user/verify-reset-otp',
        data: {
          'nicNo': nicNo,
          'email': email,
          'otp': otp,
        },
      );
      return response.statusCode == 200 || response.statusCode == 201;
    } on DioException {
      return false;
    } catch (e) {
      return false;
    }
  }

  static Future<bool> resendResetOtp(String nicNo, String email) async {
    try {
      final response = await ApiService.dio.post(
        '/auth/user/resend-reset-otp',
        data: {
          'nicNo': nicNo,
          'email': email,
        },
      );
      return response.statusCode == 200 || response.statusCode == 201;
    } on DioException {
      return false;
    } catch (e) {
      return false;
    }
  }

  static Future<bool> resetPassword(
      String nicNo, String email, String otp, String newPassword) async {
    try {
      final response = await ApiService.dio.post(
        '/auth/user/reset-password',
        data: {
          'nicNo': nicNo,
          'email': email,
          'otp': otp,
          'newPassword': newPassword,
        },
      );
      return response.statusCode == 200 || response.statusCode == 201;
    } on DioException {
      return false;
    } catch (e) {
      return false;
    }
  }

  static Future<Map<String, dynamic>> biometricLogin(
      String nicNo, String deviceId) async {
    try {
      final response = await ApiService.dio.post(
        '/auth/user/biometric-login',
        data: {
          'nicNo': nicNo,
          'deviceId': deviceId,
        },
      );
      if (response.statusCode == 200 || response.statusCode == 201) {
        final token = response.data['accessToken'];
        await SecureStorage.saveToken(token);
        return {'success': true};
      }
      return {'success': false};
    } on DioException catch (e) {
      if (e.response?.statusCode == 403 &&
          e.response?.data['code'] == 'DEVICE_MISMATCH') {
        return {
          'success': false,
          'isDeviceMismatch': true,
          'email': e.response?.data['email'] ?? '',
        };
      }
      return {'success': false};
    } catch (e) {
      return {'success': false};
    }
  }

  static Future<Map<String, dynamic>> verifyNewDevice(
      String nicNo, String deviceId, String otp) async {
    try {
      final response = await ApiService.dio.post(
        '/auth/user/verify-device',
        data: {
          'nicNo': nicNo,
          'deviceId': deviceId,
          'otp': otp,
        },
      );
      if (response.statusCode == 200 || response.statusCode == 201) {
        final token = response.data['accessToken'];
        await SecureStorage.saveToken(token);
        await SecureStorage.saveNic(nicNo);
        return {'success': true};
      }
      return {'success': false, 'message': 'Verification failed.'};
    } on DioException catch (e) {
      String errorMsg = 'Invalid OTP. Please try again.';
      if (e.response?.data != null && e.response?.data['message'] != null) {
        errorMsg = e.response?.data['message'] is List
            ? e.response?.data['message'][0]
            : e.response?.data['message'];
      }
      return {'success': false, 'message': errorMsg};
    } catch (e) {
      return {
        'success': false,
        'message': 'An error occurred during verification.'
      };
    }
  }

  static Future<bool> resendDeviceOtp(String nicNo, String email) async {
    try {
      final response = await ApiService.dio.post(
        '/auth/user/resend-device-otp',
        data: {
          'nicNo': nicNo,
          'email': email,
        },
      );
      return response.statusCode == 200 || response.statusCode == 201;
    } on DioException {
      return false;
    } catch (e) {
      return false;
    }
  }

  static Future<bool> resendHeadOtp(String username, String email) async {
    try {
      final response = await ApiService.dio.post(
        '/auth/head/resend-otp',
        data: {
          'username': username,
          'email': email,
        },
      );
      return response.statusCode == 200 || response.statusCode == 201;
    } on DioException {
      return false;
    } catch (e) {
      return false;
    }
  }

  static Future<bool> resendOfficerOtp(String badgeNo, String email) async {
    try {
      final response = await ApiService.dio.post(
        '/auth/officer/resend-otp',
        data: {
          'badgeNo': badgeNo,
          'email': email,
        },
      );
      return response.statusCode == 200 || response.statusCode == 201;
    } on DioException {
      return false;
    } catch (e) {
      return false;
    }
  }

  static Future<bool> headForgotPasswordRequest(
      String username, String email) async {
    try {
      final response = await ApiService.dio.post(
        '/auth/head/forgot-password-request',
        data: {
          'username': username,
          'email': email,
        },
      );
      return response.statusCode == 200 || response.statusCode == 201;
    } on DioException {
      return false;
    } catch (e) {
      return false;
    }
  }

  static Future<bool> headVerifyResetOtp(
      String username, String email, String otp) async {
    try {
      final response = await ApiService.dio.post(
        '/auth/head/verify-reset-otp',
        data: {
          'username': username,
          'email': email,
          'otp': otp,
        },
      );
      return response.statusCode == 200 || response.statusCode == 201;
    } on DioException {
      return false;
    } catch (e) {
      return false;
    }
  }

  static Future<bool> headResetPassword(
      String username, String email, String otp, String newPassword) async {
    try {
      final response = await ApiService.dio.post(
        '/auth/head/reset-password',
        data: {
          'username': username,
          'email': email,
          'otp': otp,
          'newPasswordStr': newPassword,
        },
      );
      return response.statusCode == 200 || response.statusCode == 201;
    } on DioException {
      return false;
    } catch (e) {
      return false;
    }
  }

  static Future<bool> officerForgotPasswordRequest(
      String badgeNo, String email) async {
    try {
      final response = await ApiService.dio.post(
        '/auth/officer/forgot-password-request',
        data: {
          'badgeNo': badgeNo,
          'email': email,
        },
      );
      return response.statusCode == 200 || response.statusCode == 201;
    } on DioException {
      return false;
    } catch (e) {
      return false;
    }
  }

  static Future<bool> officerVerifyResetOtp(
      String badgeNo, String email, String otp) async {
    try {
      final response = await ApiService.dio.post(
        '/auth/officer/verify-reset-otp',
        data: {
          'badgeNo': badgeNo,
          'email': email,
          'otp': otp,
        },
      );
      return response.statusCode == 200 || response.statusCode == 201;
    } on DioException {
      return false;
    } catch (e) {
      return false;
    }
  }

  static Future<bool> officerResetPassword(
      String badgeNo, String email, String otp, String newPassword) async {
    try {
      final response = await ApiService.dio.post(
        '/auth/officer/reset-password',
        data: {
          'badgeNo': badgeNo,
          'email': email,
          'otp': otp,
          'newPasswordStr': newPassword,
        },
      );
      return response.statusCode == 200 || response.statusCode == 201;
    } on DioException {
      return false;
    } catch (e) {
      return false;
    }
  }

  static Future<void> logout() async {
    await SecureStorage.deleteToken();
  }
}

File: backend/nest-api/src/auth/auth.service.ts
import {
  Injectable,
  UnauthorizedException,
  BadRequestException,
  NotFoundException,
  ForbiddenException,
} from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import { JwtService } from '@nestjs/jwt';
import {
  User,
  DMT_Admin,
  Police_Admin,
  Divisional_Head,
  Traffic_Officer,
} from '@prisma/client';
import * as bcrypt from 'bcrypt';
import * as nodemailer from 'nodemailer';
import { ChangePasswordDto } from './auth.controller';

export interface RegisterData {
  nicNo: string;
  name: string;
  email: string;
  password: string;
  deviceId: string;
}

@Injectable()
export class AuthService {
  constructor(
    private prisma: PrismaService,
    private jwtService: JwtService,
  ) {}

  private async sendOtpEmail(
    email: string,
    otp: string,
    type: 'registration' | 'reset',
  ) {
    const transporter = nodemailer.createTransport({
      service: 'gmail',
      auth: {
        user: process.env.EMAIL_USER,
        pass: process.env.EMAIL_PASS,
      },
    });

    let subject = '';
    let text = '';
    let html = '';

    if (type === 'registration') {
      subject = '✅ Auto-Ledger: Verify Your Email Address';
      text = `Welcome to Auto-Ledger!\n\nYour verification OTP is: ${otp}\n\nThis code will expire in 5 minutes.\n\nPlease enter this OTP in the app to complete your registration.\n\nIf you didn't register, please ignore this email.`;
      html = `
        <div style="font-family: Arial, sans-serif; max-width: 600px; margin: 0 auto; padding: 20px; background-color: #0B0F19; color: #ffffff; border-radius: 12px;">
          <h2 style="color: #00bcd4; text-align: center;">✅ Auto-Ledger</h2>
          <h3 style="text-align: center;">Verify Your Email Address</h3>
          <p style="text-align: center; color: #cccccc;">Welcome to Auto-Ledger! Please verify your email address to complete registration.</p>
          <div style="background-color: #1a1f2e; padding: 20px; border-radius: 8px; text-align: center; margin: 20px 0;">
            <h1 style="font-size: 48px; letter-spacing: 8px; color: #00bcd4; margin: 0;">${otp}</h1>
          </div>
          <p style="text-align: center; color: #aaaaaa;">This OTP is valid for <strong>5 minutes</strong>.</p>
          <hr style="border-color: #333;">
          <p style="text-align: center; color: #666666; font-size: 12px;">If you didn't request this, please ignore this email.</p>
          <p style="text-align: center; color: #666666; font-size: 12px;">© 2026 Auto-Ledger</p>
        </div>
      `;
    } else {
      subject = '🔑 Auto-Ledger: Password Reset OTP';
      text = `You requested to reset your Auto-Ledger password.\n\nYour password reset OTP is: ${otp}\n\nThis code will expire in 5 minutes.\n\nIf you didn't request a password reset, please ignore this email.`;
      html = `
        <div style="font-family: Arial, sans-serif; max-width: 600px; margin: 0 auto; padding: 20px; background-color: #0B0F19; color: #ffffff; border-radius: 12px;">
          <h2 style="color: #ff6f00; text-align: center;">🔑 Auto-Ledger</h2>
          <h3 style="text-align: center;">Password Reset Request</h3>
          <p style="text-align: center; color: #cccccc;">You requested to reset your Auto-Ledger password.</p>
          <div style="background-color: #1a1f2e; padding: 20px; border-radius: 8px; text-align: center; margin: 20px 0;">
            <h1 style="font-size: 48px; letter-spacing: 8px; color: #ff6f00; margin: 0;">${otp}</h1>
          </div>
          <p style="text-align: center; color: #aaaaaa;">This OTP is valid for <strong>5 minutes</strong>.</p>
          <hr style="border-color: #333;">
          <p style="text-align: center; color: #666666; font-size: 12px;">If you didn't request this, please ignore this email.</p>
          <p style="text-align: center; color: #666666; font-size: 12px;">© 2026 Auto-Ledger</p>
        </div>
      `;
    }

    const mailOptions = {
      from: `"Auto-Ledger" <${process.env.EMAIL_USER}>`,
      to: email,
      subject: subject,
      text: text,
      html: html,
    };

    await transporter.sendMail(mailOptions);
  }

  private generateOtp(): string {
    return Math.floor(100000 + Math.random() * 900000).toString();
  }

  async loginAdmin(username: string, pass: string, type: 'DMT' | 'POLICE') {
    let adminObj: DMT_Admin | Police_Admin | null = null;
    let roleName = '';

    if (type === 'DMT') {
      adminObj = await this.prisma.dMT_Admin.findUnique({
        where: { username: username },
      });
      roleName = 'DMT_ADMIN';
    } else {
      adminObj = await this.prisma.police_Admin.findUnique({
        where: { username: username },
      });
      roleName = 'POLICE_ADMIN';
    }

    if (!adminObj)
      throw new UnauthorizedException('Invalid Admin Username or password.');

    const isPasswordValid = await bcrypt.compare(pass, adminObj.password);
    if (!isPasswordValid)
      throw new UnauthorizedException('Invalid Admin Username or password.');

    const adminIdValue =
      type === 'DMT'
        ? (adminObj as DMT_Admin).dmt_Admin_Id
        : (adminObj as Police_Admin).police_Admin_Id;

    const payload = { sub: adminIdValue, role: roleName };
    return {
      accessToken: this.jwtService.sign(payload),
      user: { id: payload.sub, name: adminObj.name, role: roleName },
    };
  }

  async loginHead(username: string, pass: string) {
    const head = await this.prisma.divisional_Head.findUnique({
      where: { username: username },
    });

    if (!head || !head.is_Active)
      throw new UnauthorizedException('Invalid or inactive Head account.');

    const isPasswordValid = await bcrypt.compare(pass, head.password);
    if (!isPasswordValid)
      throw new UnauthorizedException('Invalid Head Username or password.');

    const payload = {
      sub: head.divisional_Head_Id,
      role: head.role,
      divisionId: head.division_Id,
    };
    return {
      accessToken: this.jwtService.sign(payload),
      user: {
        id: head.divisional_Head_Id,
        name: head.name,
        email: head.email,
        role: head.role,
        divisionId: head.division_Id,
      },
    };
  }

  async loginOfficer(badgeNo: string, pass: string) {
    const officer = await this.prisma.traffic_Officer.findUnique({
      where: { badge_No: badgeNo },
      include: { shifts: true },
    });

    if (!officer)
      throw new UnauthorizedException('Invalid Badge Number or password.');

    const isPasswordValid = await bcrypt.compare(pass, officer.password);
    if (!isPasswordValid)
      throw new UnauthorizedException('Invalid Badge Number or password.');

    const now = new Date();
    const activeShift = officer.shifts.find(
      (shift) =>
        shift.is_Active &&
        new Date(shift.start_Time) <= now &&
        new Date(shift.end_Time) >= now,
    );

    if (!activeShift) {
      throw new ForbiddenException(
        'Access Denied: You are not within an active shift schedule.',
      );
    }

    const payload = {
      sub: officer.traffic_Officer_Id,
      role: officer.role,
      badgeNo: officer.badge_No,
      headId: officer.divisional_Head_Id,
    };
    return {
      accessToken: this.jwtService.sign(payload),
      user: {
        id: officer.traffic_Officer_Id,
        name: officer.name,
        email: officer.email,
        role: officer.role,
        badgeNo: officer.badge_No,
      },
    };
  }

  async changePassword(userId: string, role: string, dto: ChangePasswordDto) {
    let user: Divisional_Head | Traffic_Officer | null = null;

    if (role === 'DIVISIONAL_HEAD') {
      user = await this.prisma.divisional_Head.findUnique({
        where: { divisional_Head_Id: userId },
      });
    } else if (role === 'TRAFFIC_OFFICER') {
      user = await this.prisma.traffic_Officer.findUnique({
        where: { traffic_Officer_Id: userId },
      });
    } else {
      throw new BadRequestException('Invalid role for password change');
    }

    if (!user) {
      throw new NotFoundException('User not found');
    }

    const isPasswordValid = await bcrypt.compare(
      dto.oldPassword,
      user.password,
    );
    if (!isPasswordValid) {
      throw new UnauthorizedException('Invalid old password');
    }

    const hashedNewPassword = await bcrypt.hash(dto.newPassword, 10);

    if (role === 'DIVISIONAL_HEAD') {
      await this.prisma.divisional_Head.update({
        where: { divisional_Head_Id: userId },
        data: { password: hashedNewPassword },
      });
    } else {
      await this.prisma.traffic_Officer.update({
        where: { traffic_Officer_Id: userId },
        data: { password: hashedNewPassword },
      });
    }

    return { message: 'Password changed successfully' };
  }

  async requestHeadPasswordReset(username: string, email: string) {
    const head = await this.prisma.divisional_Head.findUnique({
      where: { username: username },
    });

    if (!head || head.email !== email || !head.is_Active) {
      throw new BadRequestException('Invalid Username or Email provided.');
    }

    const otp = this.generateOtp();
    const expiresAt = new Date(Date.now() + 5 * 60 * 1000);

    await this.prisma.divisional_Head.update({
      where: { username: username },
      data: { reset_Otp: otp, reset_Otp_Expires_At: expiresAt },
    });

    await this.sendOtpEmail(email, otp, 'reset');
    return { message: 'OTP sent successfully to your email.' };
  }

  async resetHeadPassword(
    username: string,
    email: string,
    otp: string,
    newPasswordStr: string,
  ) {
    const head = await this.prisma.divisional_Head.findUnique({
      where: { username: username },
    });

    if (!head || head.email !== email || head.reset_Otp !== otp) {
      throw new BadRequestException('Invalid OTP or Credentials.');
    }

    if (!head.reset_Otp_Expires_At || new Date() > head.reset_Otp_Expires_At) {
      throw new BadRequestException('OTP has expired.');
    }

    const hashedPassword = await bcrypt.hash(newPasswordStr, 10);

    await this.prisma.divisional_Head.update({
      where: { username: username },
      data: {
        password: hashedPassword,
        reset_Otp: null,
        reset_Otp_Expires_At: null,
      },
    });

    return { message: 'Divisional Head password reset successfully.' };
  }

  async requestOfficerPasswordReset(badgeNo: string, email: string) {
    const officer = await this.prisma.traffic_Officer.findUnique({
      where: { badge_No: badgeNo },
    });

    if (!officer || officer.email !== email) {
      throw new BadRequestException('Invalid Badge Number or Email provided.');
    }

    const otp = this.generateOtp();
    const expiresAt = new Date(Date.now() + 5 * 60 * 1000);

    await this.prisma.traffic_Officer.update({
      where: { badge_No: badgeNo },
      data: { reset_Otp: otp, reset_Otp_Expires_At: expiresAt },
    });

    await this.sendOtpEmail(email, otp, 'reset');
    return { message: 'OTP sent successfully to your email.' };
  }

  async resetOfficerPassword(
    badgeNo: string,
    email: string,
    otp: string,
    newPasswordStr: string,
  ) {
    const officer = await this.prisma.traffic_Officer.findUnique({
      where: { badge_No: badgeNo },
    });

    if (!officer || officer.email !== email || officer.reset_Otp !== otp) {
      throw new BadRequestException('Invalid OTP or Credentials.');
    }

    if (
      !officer.reset_Otp_Expires_At ||
      new Date() > officer.reset_Otp_Expires_At
    ) {
      throw new BadRequestException('OTP has expired.');
    }

    const hashedPassword = await bcrypt.hash(newPasswordStr, 10);

    await this.prisma.traffic_Officer.update({
      where: { badge_No: badgeNo },
      data: {
        password: hashedPassword,
        reset_Otp: null,
        reset_Otp_Expires_At: null,
      },
    });

    return { message: 'Traffic Officer password reset successfully.' };
  }

  private generateUserToken(user: User) {
    const payload = { sub: user.user_Id, nic: user.nic_No, role: 'USER' };
    return {
      accessToken: this.jwtService.sign(payload),
      user: {
        id: user.user_Id,
        name: user.name,
        nic: user.nic_No,
        email: user.email,
        isEmailVerified: user.isEmailVerified,
      },
    };
  }

  async registerUser(data: RegisterData) {
    const user = await this.prisma.user.findUnique({
      where: { nic_No: data.nicNo },
    });

    if (!user) {
      throw new BadRequestException(
        'Registration Failed: No user found for this NIC.',
      );
    }

    const license = await this.prisma.driving_License.findUnique({
      where: { user_Id: user.user_Id },
    });

    if (!license) {
      throw new BadRequestException(
        'Registration Failed: No driving license found for this NIC.',
      );
    }

    const hashedPassword = await bcrypt.hash(data.password, 10);
    const otp = this.generateOtp();
    const expiresAt = new Date(Date.now() + 5 * 60 * 1000);

    await this.prisma.user.update({
      where: { nic_No: data.nicNo },
      data: {
        name: data.name,
        email: data.email,
        password: hashedPassword,
        device_Id: data.deviceId,
        isEmailVerified: false,
        reset_Otp: otp,
        reset_Otp_Expires_At: expiresAt,
      },
    });

    await this.sendOtpEmail(data.email, otp, 'registration');

    return {
      message: 'OTP sent to your email. Please verify.',
      success: true,
    };
  }

  async verifyRegistration(nicNo: string, otp: string) {
    const user = await this.prisma.user.findUnique({
      where: { nic_No: nicNo },
    });

    if (!user) throw new BadRequestException('User not found.');
    if (user.reset_Otp !== otp) throw new BadRequestException('Invalid OTP.');
    if (!user.reset_Otp_Expires_At || new Date() > user.reset_Otp_Expires_At) {
      throw new BadRequestException('OTP has expired.');
    }

    const updatedUser = await this.prisma.user.update({
      where: { nic_No: nicNo },
      data: {
        isEmailVerified: true,
        reset_Otp: null,
        reset_Otp_Expires_At: null,
      },
    });

    return this.generateUserToken(updatedUser);
  }

  async loginUser(nicNo: string, pass: string, deviceId: string) {
    const user = await this.prisma.user.findUnique({
      where: { nic_No: nicNo },
    });

    if (!user) throw new UnauthorizedException('Invalid NIC or password.');

    if (!user.isEmailVerified) {
      throw new ForbiddenException('Please verify your email using OTP first.');
    }

    const isPasswordValid = await bcrypt.compare(pass, user.password);
    if (!isPasswordValid)
      throw new UnauthorizedException('Invalid NIC or password.');

    if (user.device_Id !== deviceId) {
      throw new ForbiddenException({
        code: 'DEVICE_MISMATCH',
        message: 'New device detected. OTP verification required.',
        email: user.email,
      });
    }

    return this.generateUserToken(user);
  }

  async verifyNewDevice(nicNo: string, newDeviceId: string) {
    const user = await this.prisma.user.update({
      where: { nic_No: nicNo },
      data: {
        device_Id: newDeviceId,
        isEmailVerified: true,
      },
    });

    return this.generateUserToken(user);
  }

  async biometricLogin(nicNo: string, deviceId: string) {
    const user = await this.prisma.user.findUnique({
      where: { nic_No: nicNo },
    });

    if (!user) throw new UnauthorizedException('Invalid user.');

    if (!user.isEmailVerified) {
      throw new ForbiddenException('Please verify your email using OTP first.');
    }

    if (user.device_Id !== deviceId) {
      throw new UnauthorizedException(
        'Biometric Access Denied: Unrecognized device.',
      );
    }

    return this.generateUserToken(user);
  }

  async changeUserPassword(
    userId: string,
    dto: { oldPassword: string; newPassword: string },
  ) {
    const user = await this.prisma.user.findUnique({
      where: { user_Id: userId },
    });

    if (!user) throw new NotFoundException('User not found');

    const isPasswordValid = await bcrypt.compare(
      dto.oldPassword,
      user.password,
    );
    if (!isPasswordValid) {
      throw new UnauthorizedException('Invalid old password');
    }

    const hashedNewPassword = await bcrypt.hash(dto.newPassword, 10);

    await this.prisma.user.update({
      where: { user_Id: userId },
      data: { password: hashedNewPassword },
    });

    return { message: 'User password changed successfully' };
  }

  async requestPasswordReset(nicNo: string, email: string) {
    const user = await this.prisma.user.findUnique({
      where: { nic_No: nicNo },
    });

    if (!user || user.email !== email) {
      throw new BadRequestException('Invalid NIC or Email provided.');
    }

    const otp = this.generateOtp();
    const expiresAt = new Date(Date.now() + 5 * 60 * 1000);

    await this.prisma.user.update({
      where: { nic_No: nicNo },
      data: { reset_Otp: otp, reset_Otp_Expires_At: expiresAt },
    });

    await this.sendOtpEmail(email, otp, 'reset');
    return { message: 'OTP sent successfully to your email.' };
  }

  async resetPassword(
    nicNo: string,
    email: string,
    otp: string,
    newPasswordStr: string,
  ) {
    const user = await this.prisma.user.findUnique({
      where: { nic_No: nicNo },
    });

    if (!user || user.email !== email) {
      throw new BadRequestException('Invalid NIC or Email.');
    }

    if (user.reset_Otp !== otp) throw new BadRequestException('Invalid OTP.');
    if (!user.reset_Otp_Expires_At || new Date() > user.reset_Otp_Expires_At) {
      throw new BadRequestException('OTP has expired.');
    }

    const hashedNewPassword = await bcrypt.hash(newPasswordStr, 10);

    await this.prisma.user.update({
      where: { nic_No: nicNo },
      data: {
        password: hashedNewPassword,
        reset_Otp: null,
        reset_Otp_Expires_At: null,
      },
    });

    return {
      message: 'Password has been reset successfully. You can now login.',
    };
  }

  async resendRegistrationOtp(nicNo: string) {
    const user = await this.prisma.user.findUnique({
      where: { nic_No: nicNo },
    });

    if (!user) {
      throw new BadRequestException('User not found.');
    }

    if (user.isEmailVerified) {
      throw new BadRequestException('Email already verified.');
    }

    const otp = this.generateOtp();
    const expiresAt = new Date(Date.now() + 5 * 60 * 1000);

    await this.prisma.user.update({
      where: { nic_No: nicNo },
      data: {
        reset_Otp: otp,
        reset_Otp_Expires_At: expiresAt,
      },
    });

    await this.sendOtpEmail(user.email, otp, 'registration');
    return {
      message: 'OTP resent successfully. Please check your email.',
      success: true,
    };
  }

  async resendResetOtp(nicNo: string, email: string) {
    const user = await this.prisma.user.findUnique({
      where: { nic_No: nicNo },
    });

    if (!user || user.email !== email) {
      throw new BadRequestException('Invalid NIC or Email.');
    }

    const otp = this.generateOtp();
    const expiresAt = new Date(Date.now() + 5 * 60 * 1000);

    await this.prisma.user.update({
      where: { nic_No: nicNo },
      data: {
        reset_Otp: otp,
        reset_Otp_Expires_At: expiresAt,
      },
    });

    await this.sendOtpEmail(email, otp, 'reset');
    return {
      message: 'OTP resent successfully. Please check your email.',
      success: true,
    };
  }

  async resendDeviceOtp(nicNo: string, email: string) {
    const user = await this.prisma.user.findUnique({
      where: { nic_No: nicNo },
    });

    if (!user || user.email !== email) {
      throw new BadRequestException('Invalid NIC or Email.');
    }

    const otp = this.generateOtp();
    const expiresAt = new Date(Date.now() + 5 * 60 * 1000);

    await this.prisma.user.update({
      where: { nic_No: nicNo },
      data: {
        reset_Otp: otp,
        reset_Otp_Expires_At: expiresAt,
      },
    });

    await this.sendOtpEmail(email, otp, 'reset');
    return {
      message: 'OTP resent successfully. Please check your email.',
      success: true,
    };
  }
}
